### PR TITLE
MessageDispatcher sorts batch per dispatch consistency and reuses the connections more efficiently

### PR DESCRIPTION
**Pull request by danielmarbach**
Connects to https://github.com/Particular/EndToEnd/issues/3
Relates to #181 

In order to understand the implications of #181 I needed to dig deeper into the Sql Transport code. While doing this I realized we are using unnecessary connections inside the same batch. This PR reuses the existing code but does the following:

It creates two paths. One for `Isolated` dispatch consistency and one for `default` consistency.

Side note: The core has a small design flaw. We are marking all operations in a batch as `Isolated` when the outbox is used. Under the outbox we would actually not required any transaction scope. The only thing we would need to make sure is that we are not participating in the receive transaction. But since the core API doesn't allow to differentiate between immediate dispatch and "outbox" mode we need to treat both consistently.

All operations which are isolated will be send inside the same transaction scope with `RequiresNew` on one newly created connection.

All operations which have default consistency will be sent either by using the connection and transaction provided by the message pump or by a newly created connection and transaction.

So in all the modes when we do not have a connection and transaction provided by the message pump we will essentially have one connection instead N connections while N being the number of messages in the batch.

@Particular/transport-maintainers please review

### CODE REVIEW

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 1]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
-            var queue = new TableBasedQueue(destination);
-
-            //Dispatch in separate transaction even if transaction scope already exists
-            if (operation.RequiredDispatchConsistency == DispatchConsistency.Isolated)
-            {
-                await DispatchInIsolatedTransactionScope(queue, operation).ConfigureAwait(false);
-
-                return;
-            }
+            var defaultConsistencyOperations = operations.UnicastTransportOperations.Where(o => o.RequiredDispatchConsistency == DispatchConsistency.Default).ToList();
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: Can we move this to upper method?

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 2]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
         // We need to check if we can support cancellation in here as well?
         public async Task Dispatch(TransportOperations operations, ContextBag context)
         {
-            foreach (var operation in operations.UnicastTransportOperations)
-            {
-                await DispatchUnicastOperation(context, operation).ConfigureAwait(false);
-            }
+            await DispatchIsolatedConsistencyTransportOperations(operations).ConfigureAwait(false);
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: Can we change name to something simpler/shorter like `DispatchAsIsolated` - assuming that we do https://github.com/Particular/NServiceBus.SqlServer/pull/198/files#r59613507

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 3]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
+
+            await DispatchDefaultConsistencyTransportOperations(operations, context).ConfigureAwait(false);
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: Same as above maybe `DispatchAsNonIsolated`?

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 4]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
+                var queue = new TableBasedQueue(destination);
+
+                await queue.SendMessage(operation.Message, sqlTransportConnection, sqlTransportTransaction).ConfigureAwait(false);
+            }
+        }
+
+        private async Task DispatchOperationsWithNewConnectionAndTransaction(List<UnicastTransportOperation> defaultConsistencyOperations)
         {
             using (var connection = await connectionFactory.OpenNewConnection().ConfigureAwait(false))
             {
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: '{' is not needed here.

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 5]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
                 using (var transaction = connection.BeginTransaction())
                 {
-                    await queue.SendMessage(operation.Message, connection, transaction).ConfigureAwait(false);
+                    foreach (var operation in defaultConsistencyOperations)
+                    {
+                        var destination = addressParser.Parse(operation.Destination);
+                        var queue = new TableBasedQueue(destination);
 
+                        await queue.SendMessage(operation.Message, connection, transaction).ConfigureAwait(false);
+                    }
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: New line after `}`

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 6]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
                     transaction.Commit();
                 }
             }
         }
 
-        async Task DispatchInCurrentTransportTransaction(TransportTransaction transportTransaction, TableBasedQueue queue, UnicastTransportOperation operation)
+        private async Task DispatchIsolatedConsistencyTransportOperations(TransportOperations operations)
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: No need for `private`

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 7]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
+            foreach (var operation in defaultConsistencyOperations)
+            {
+                var destination = addressParser.Parse(operation.Destination);
+                var queue = new TableBasedQueue(destination);
+
+                await queue.SendMessage(operation.Message, sqlTransportConnection, sqlTransportTransaction).ConfigureAwait(false);
+            }
+        }
+
+        private async Task DispatchOperationsWithNewConnectionAndTransaction(List<UnicastTransportOperation> defaultConsistencyOperations)
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: No need for `private`

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: And pass into the method? Could do, my taste was on the upper level the properly named high level entry points and on the level of usage the where filtering. When your approach pleases more the overall approach here, then happy to change

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 8]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
+                await DispatchOperationsWithNewConnectionAndTransaction(defaultConsistencyOperations).ConfigureAwait(false);
                 return;
             }
 
-            await DispatchInCurrentTransportTransaction(transportTransaction, queue, operation).ConfigureAwait(false);
+            await DispatchReusingTheReceiveConnectionAndTransaction(transportTransaction, defaultConsistencyOperations).ConfigureAwait(false);
         }
 
-        async Task DispatchAsSeparateSendOperation(TableBasedQueue queue, UnicastTransportOperation operation)
+        private async Task DispatchReusingTheReceiveConnectionAndTransaction(TransportTransaction transportTransaction, List<UnicastTransportOperation> defaultConsistencyOperations)
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: No need for `private`

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs [hunk 9]
```diff
@@ -16,83 +18,92 @@ public MessageDispatcher(SqlConnectionFactory connectionFactory, QueueAddressPar
...
             if (transportTransactionExists == false || InReceiveOnlyTransportTransactionMode(transportTransaction))
             {
-                await DispatchAsSeparateSendOperation(queue, operation).ConfigureAwait(false);
-
+                await DispatchOperationsWithNewConnectionAndTransaction(defaultConsistencyOperations).ConfigureAwait(false);
                 return;
             }
 
-            await DispatchInCurrentTransportTransaction(transportTransaction, queue, operation).ConfigureAwait(false);
+            await DispatchReusingTheReceiveConnectionAndTransaction(transportTransaction, defaultConsistencyOperations).ConfigureAwait(false);
```
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: Maybe `DispatchUsingReceiveTransaction` would be better name?

**Comment by tmasternak** [comment]
@danielmarbach done with review

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 6 above)
Comment (danielmarbach) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: Then this repos cleanup is out of date. Ran Ctr+E+F

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 7 above)
Comment (danielmarbach) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: See above

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 1 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: I like the idea of high level entry points my only argument would be that it's not obvious what happens to the collection passed into it. Now when I look at the name `DispatchDefaultConsistencyTransportOperations` it does suggests that the filtering is done - so I'm good with keeping the filtering inside the methods.

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 6 above)
Comment (tmasternak) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: Yes something is off. I'm looking into it.

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 3 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: how about
```
var default =  operations.UnicastTransportOperations.Where(o => o.RequiredDispatchConsistency == DispatchConsistency.Default);
var isolated =  operations.UnicastTransportOperations.Where(o => o.RequiredDispatchConsistency == DispatchConsistency.Default);
DispatchIsolated(isolated);
DispatchNonIsolated(default);
```

I mean, when passing the same collection into two methods, from the caller perspective it is not obvious if both methods wont take the same operations etc. With that code we are explicitly categorizing the operations before passing them to proper dispatch methods.

File: src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer/Sending/MessageDispatcher.cs]: LoL, I've just realized I wrote the same comment above ;-)

**Comment by danielmarbach** [comment]
@tmasternak updated

**Comment by tmasternak** [comment]
@danielmarbach Looks good to me. Let's wait for green build and I'll merge.

**Comment by danielmarbach** [comment]
Green

**Comment by tmasternak** [comment]
@danielmarbach good stuff!

**Comment by SzymonPobiega** [comment]
Awesome teamwork

**Pull request merged**